In [95]:
import glob

import pandas as pd
import numpy as np

In [96]:
# Initialise a dataframe to store data
bike_df = pd.DataFrame()

# List of tripdata csv file pathnames, that are concatenated into bike dataframe
data_list = [str(i) for i in glob.glob('data\JC-2016??-citibike-tripdata.csv')]
bike_df = pd.concat([pd.read_csv(file) for file in data_list])

# Initial cleaning and reordering of columns
bike_df.reset_index(drop=True, inplace=True)
bike_df['ID'] = bike_df.index 
bike_df = bike_df[['ID', 'Trip Duration', 'Start Time', 'Start Station ID',
       'Start Station Name', 'Start Station Latitude',
       'Start Station Longitude', 'Stop Time', 'End Station ID', 'End Station Name',
       'End Station Latitude', 'End Station Longitude', 'Bike ID', 'User Type',
       'Birth Year', 'Gender']]

In [97]:
# Initial Inspect
bike_df.head(10)

ID  Trip Duration           Start Time  Start Station ID  \
0   0            362  2016-01-01 00:02:52              3186   
1   1            200  2016-01-01 00:18:22              3186   
2   2            202  2016-01-01 00:18:25              3186   
3   3            248  2016-01-01 00:23:13              3209   
4   4            903  2016-01-01 01:03:20              3195   
5   5            883  2016-01-01 01:03:28              3195   
6   6            445  2016-01-01 01:07:45              3186   
7   7            192  2016-01-01 01:18:51              3211   
8   8            409  2016-01-01 01:23:44              3187   
9   9            285  2016-01-01 01:25:12              3187   

  Start Station Name  Start Station Latitude  Start Station Longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   
5            Sip Ave               40.730743               -74.063784   
6      Grove St PATH               40.719586               -74.043117   
7         Newark Ave               40.721525               -74.046305   
8          Warren St               40.721124               -74.038051   
9          Warren St               40.721124               -74.038051   

             Stop Time  End Station ID  End Station Name  \
0  2016-01-01 00:08:54            3209      Brunswick St   
1  2016-01-01 00:21:42            3213    Van Vorst Park   
2  2016-01-01 00:21:47            3213    Van Vorst Park   
3  2016-01-01 00:27:21            3203     Hamilton Park   
4  2016-01-01 01:18:24            3210    Pershing Field   
5  2016-01-01 01:18:11            3210    Pershing Field   
6  2016-01-01 01:15:11            3203     Hamilton Park   
7  2016-01-01 01:22:03            3203     Hamilton Park   
8  2016-01-01 01:30:34            3214  Essex Light Rail   
9  2016-01-01 01:29:57            3214  Essex Light Rail   

   End Station Latitude  End Station Longitude  Bike ID   User Type  \
0             40.724176             -74.050656    24647  Subscriber   
1             40.718489             -74.047727    24605  Subscriber   
2             40.718489             -74.047727    24689  Subscriber   
3             40.727596             -74.044247    24693  Subscriber   
4             40.742677             -74.051789    24573    Customer   
5             40.742677             -74.051789    24442    Customer   
6             40.727596             -74.044247    24510  Subscriber   
7             40.727596             -74.044247    24625  Subscriber   
8             40.712774             -74.036486    24429  Subscriber   
9             40.712774             -74.036486    24407  Subscriber   

   Birth Year  Gender  
0      1964.0       2  
1      1962.0       1  
2      1962.0       2  
3      1984.0       1  
4         NaN       0  
5         NaN       0  
6      1988.0       2  
7      1980.0       1  
8      1990.0       1  
9      1988.0       2

In [98]:
bike_df.tail(10)

ID  Trip Duration           Start Time  Start Station ID  \
247574  247574            216  2016-12-31 21:48:33              3205   
247575  247575            559  2016-12-31 21:52:13              3193   
247576  247576            473  2016-12-31 22:04:14              3186   
247577  247577            242  2016-12-31 23:02:01              3214   
247578  247578            377  2016-12-31 23:10:00              3206   
247579  247579            557  2016-12-31 23:10:16              3214   
247580  247580           2749  2016-12-31 23:29:39              3183   
247581  247581            173  2016-12-31 23:44:37              3186   
247582  247582           2424  2016-12-31 23:44:50              3214   
247583  247583           2419  2016-12-31 23:44:50              3214   

       Start Station Name  Start Station Latitude  Start Station Longitude  \
247574  JC Medical Center               40.716540               -74.049638   
247575       Lincoln Park               40.724605               -74.078406   
247576      Grove St PATH               40.719586               -74.043117   
247577   Essex Light Rail               40.712774               -74.036486   
247578            Hilltop               40.731169               -74.057574   
247579   Essex Light Rail               40.712774               -74.036486   
247580     Exchange Place               40.716247               -74.033459   
247581      Grove St PATH               40.719586               -74.043117   
247582   Essex Light Rail               40.712774               -74.036486   
247583   Essex Light Rail               40.712774               -74.036486   

                  Stop Time  End Station ID       End Station Name  \
247574  2016-12-31 21:52:09            3186          Grove St PATH   
247575  2016-12-31 22:01:32            3195                Sip Ave   
247576  2016-12-31 22:12:07            3268         Lafayette Park   
247577  2016-12-31 23:06:03            3276       Marin Light Rail   
247578  2016-12-31 23:16:17            3225  Baldwin at Montgomery   
247579  2016-12-31 23:19:33            3203          Hamilton Park   
247580  2017-01-01 00:15:29            3183         Exchange Place   
247581  2016-12-31 23:47:31            3270        Jersey & 6th St   
247582  2017-01-01 00:25:14            3214       Essex Light Rail   
247583  2017-01-01 00:25:10            3214       Essex Light Rail   

        End Station Latitude  End Station Longitude  Bike ID   User Type  \
247574             40.719586             -74.043117    24590  Subscriber   
247575             40.730743             -74.063784    26431  Subscriber   
247576             40.713464             -74.062859    24525  Subscriber   
247577             40.714584             -74.042817    26232  Subscriber   
247578             40.723659             -74.064194    24421  Subscriber   
247579             40.727596             -74.044247    24465  Subscriber   
247580             40.716247             -74.033459    24389    Customer   
247581             40.725289             -74.045572    24641  Subscriber   
247582             40.712774             -74.036486    26219  Subscriber   
247583             40.712774             -74.036486    24471  Subscriber   

        Birth Year  Gender  
247574      1985.0       1  
247575      1986.0       1  
247576      1991.0       1  
247577      1958.0       1  
247578      1984.0       1  
247579      1981.0       2  
247580         NaN       0  
247581      1978.0       1  
247582      1960.0       2  
247583      1956.0       1

In [99]:
bike_df.describe()

ID  Trip Duration  Start Station ID  Start Station Latitude  \
count  247584.000000   2.475840e+05     247584.000000           247584.000000   
mean   123791.500000   8.856305e+02       3207.065206               40.723121   
std     71471.488861   3.593798e+04         26.955103                0.008199   
min         0.000000   6.100000e+01       3183.000000               40.692640   
25%     61895.750000   2.480000e+02       3186.000000               40.717732   
50%    123791.500000   3.900000e+02       3201.000000               40.721525   
75%    185687.250000   6.660000e+02       3211.000000               40.727596   
max    247583.000000   1.632981e+07       3426.000000               40.752559   

       Start Station Longitude  End Station ID  End Station Latitude  \
count            247584.000000   247584.000000         247584.000000   
mean                -74.046438     3203.572553             40.722594   
std                   0.011211       61.579494              0.007958   
min                 -74.096937      147.000000             40.692216   
25%                 -74.050656     3186.000000             40.716540   
50%                 -74.044247     3199.000000             40.721124   
75%                 -74.038051     3211.000000             40.727224   
max                 -74.032108     3426.000000             40.801343   

       End Station Longitude        Bike ID     Birth Year         Gender  
count          247584.000000  247584.000000  228585.000000  247584.000000  
mean              -74.045855   24935.260481    1979.335276       1.123534  
std                 0.011283     748.469712       9.596809       0.518687  
min               -74.096937   14552.000000    1900.000000       0.000000  
25%               -74.050444   24491.000000    1974.000000       1.000000  
50%               -74.043117   24609.000000    1981.000000       1.000000  
75%               -74.036486   24719.000000    1986.000000       1.000000  
max               -73.957390   27274.000000    2000.000000       2.000000

In [100]:
# Searching for duplicated rows
print('Duplicated Rows:')
print(bike_df.duplicated().sum())

# Data Integrity Check
print('\n')
print('Missing values per variable:')
bike_df.isna().sum()

Duplicated Rows:
0


Missing values per variable:


ID                             0
Trip Duration                  0
Start Time                     0
Start Station ID               0
Start Station Name             0
Start Station Latitude         0
Start Station Longitude        0
Stop Time                      0
End Station ID                 0
End Station Name               0
End Station Latitude           0
End Station Longitude          0
Bike ID                        0
User Type                    380
Birth Year                 18999
Gender                         0
dtype: int64

In [101]:
bike_df.dtypes

ID                           int64
Trip Duration                int64
Start Time                  object
Start Station ID             int64
Start Station Name          object
Start Station Latitude     float64
Start Station Longitude    float64
Stop Time                   object
End Station ID               int64
End Station Name            object
End Station Latitude       float64
End Station Longitude      float64
Bike ID                      int64
User Type                   object
Birth Year                 float64
Gender                       int64
dtype: object

- Start and Stop times should be converted to DATETIME types.

In [102]:
bike_df['Start Time'] = pd.to_datetime(bike_df['Start Time'])
bike_df['Stop Time'] = pd.to_datetime(bike_df['Stop Time'])
bike_df.dtypes

ID                                  int64
Trip Duration                       int64
Start Time                 datetime64[ns]
Start Station ID                    int64
Start Station Name                 object
Start Station Latitude            float64
Start Station Longitude           float64
Stop Time                  datetime64[ns]
End Station ID                      int64
End Station Name                   object
End Station Latitude              float64
End Station Longitude             float64
Bike ID                             int64
User Type                          object
Birth Year                        float64
Gender                              int64
dtype: object

In [111]:
# Remove spaces from column names
bike_df.columns = [x.replace(' ', '_').lower() for x in bike_df.columns]
print(bike_df.columns)

Index(['id', 'trip_duration', 'start_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'stop_time', 'end_station_id',
       'end_station_name', 'end_station_latitude', 'end_station_longitude',
       'bike_id', 'user_type', 'birth_year', 'gender', 'age'],
      dtype='object')


### Found Issues to investigate
- User type has 380 missing values.
- Birth Year has a unexpectadly small minimum value (1900) and 19 thousand missing values
- Trip duration has an extremely high max value (5 orders of magnitude greater than it's 75th percentile value)
- Gender does not appear to be missing data but values of 0 indicated an unknown gender, these might need to be removed for gender based analysis/ dealt with.

In [119]:
# Find list of lowest birth years
bike_df['age'] = 2016 - bike_df.birth_year.values
bike_df[['id', 'birth_year', 'age']].sort_values('age', ascending=False)[:10]



id  birth_year    age
215055  215055      1900.0  116.0
136211  136211      1934.0   82.0
164274  164274      1937.0   79.0
187190  187190      1937.0   79.0
180373  180373      1937.0   79.0
213567  213567      1937.0   79.0
177320  177320      1940.0   76.0
112336  112336      1940.0   76.0
149803  149803      1940.0   76.0
124501  124501      1941.0   75.0

- There seems to be a single outlier birth year of 1900 (age = 116) under index 215055 that can be dropped
- It will be unexpected to encounter further riders of that age, so the observation can be safely dropped

In [120]:
# Drop observation in place and inspect the new top ten highest age rider records
bike_df.drop([215055], inplace=True)
bike_df[['id', 'birth_year', 'age']].sort_values('age', ascending=False)[:10]

id  birth_year   age
136211  136211      1934.0  82.0
187190  187190      1937.0  79.0
164274  164274      1937.0  79.0
213567  213567      1937.0  79.0
180373  180373      1937.0  79.0
112336  112336      1940.0  76.0
149803  149803      1940.0  76.0
177320  177320      1940.0  76.0
242203  242203      1941.0  75.0
208908  208908      1941.0  75.0

In [130]:
# First 20 missing birth year records
missing_birth_year_observations = bike_df[bike_df.birth_year.isnull()]
missing_birth_year_observations.head(20)

id  trip_duration          start_time  start_station_id  \
4      4            903 2016-01-01 01:03:20              3195   
5      5            883 2016-01-01 01:03:28              3195   
22    22            988 2016-01-01 03:16:33              3196   
53    53           3090 2016-01-01 11:07:15              3203   
57    57            788 2016-01-01 11:50:30              3210   
58    58            767 2016-01-01 11:50:43              3210   
59    59            682 2016-01-01 11:55:01              3193   
61    61           1886 2016-01-01 12:02:58              3192   
63    63           1667 2016-01-01 12:40:57              3199   
67    67           1410 2016-01-01 13:02:11              3199   
69    69            657 2016-01-01 13:14:30              3193   
75    75            828 2016-01-01 13:32:45              3186   
140  140            485 2016-01-01 19:07:15              3199   
144  144            167 2016-01-01 19:44:53              3202   
165  165            865 2016-01-02 01:32:14              3195   
166  166            836 2016-01-02 01:32:24              3195   
205  205           1536 2016-01-02 11:19:09              3199   
206  206           1524 2016-01-02 11:19:27              3199   
207  207           1036 2016-01-02 11:26:11              3183   
208  208            989 2016-01-02 11:26:12              3183   

     start_station_name  start_station_latitude  start_station_longitude  \
4               Sip Ave               40.730743               -74.063784   
5               Sip Ave               40.730743               -74.063784   
22       Riverview Park               40.744319               -74.043991   
53        Hamilton Park               40.727596               -74.044247   
57       Pershing Field               40.742677               -74.051789   
58       Pershing Field               40.742677               -74.051789   
59         Lincoln Park               40.724605               -74.078406   
61   Liberty Light Rail               40.711242               -74.055701   
63         Newport Pkwy               40.728745               -74.032108   
67         Newport Pkwy               40.728745               -74.032108   
69         Lincoln Park               40.724605               -74.078406   
75        Grove St PATH               40.719586               -74.043117   
140        Newport Pkwy               40.728745               -74.032108   
144        Newport PATH               40.727224               -74.033759   
165             Sip Ave               40.730743               -74.063784   
166             Sip Ave               40.730743               -74.063784   
205        Newport Pkwy               40.728745               -74.032108   
206        Newport Pkwy               40.728745               -74.032108   
207      Exchange Place               40.716247               -74.033459   
208      Exchange Place               40.716247               -74.033459   

              stop_time  end_station_id    end_station_name  \
4   2016-01-01 01:18:24            3210      Pershing Field   
5   2016-01-01 01:18:11            3210      Pershing Field   
22  2016-01-01 03:33:02            3209        Brunswick St   
53  2016-01-01 11:58:46            3203       Hamilton Park   
57  2016-01-01 12:03:39            3195             Sip Ave   
58  2016-01-01 12:03:30            3195             Sip Ave   
59  2016-01-01 12:06:24            3193        Lincoln Park   
61  2016-01-01 12:34:25            3192  Liberty Light Rail   
63  2016-01-01 13:08:44            3199        Newport Pkwy   
67  2016-01-01 13:25:41            3185           City Hall   
69  2016-01-01 13:25:28            3193        Lincoln Park   
75  2016-01-01 13:46:33            3199        Newport Pkwy   
140 2016-01-01 19:15:21            3187           Warren St   
144 2016-01-01 19:47:40            3199        Newport Pkwy   
165 2016-01-02 01:46:39            3210      Pershing Field   
166 2016-01-02 01:46:20            

#### Observations
- Gender seems to always be 0 (unknown)
- User type seems to always be 'Customer'


In [150]:
print("Missing birth year obvservations by Customer vs Subscriber 'user_type': ")
print(missing_birth_year_observations.user_type.value_counts(dropna=False))
print('\n')


print("Percentage of each user type with missing birth year observation:")
print(missing_birth_year_observations.user_type.value_counts() / bike_df.user_type.value_counts())
print('\n')

print('Number of missing birth years per gender type:')
print(missing_birth_year_observations.gender.value_counts())

Missing birth year obvservations by Customer vs Subscriber 'user_type': 
Customer      15470
Subscriber     3529
Name: user_type, dtype: int64


Percentage of each user type with missing birth year observation:
Customer      0.996714
Subscriber    0.015232
Name: user_type, dtype: float64


Number of missing birth years per gender type:
0    18999
Name: gender, dtype: int64


- The majority of missing birth year observations are from Customers ('Customer' user_type)
    - 99.67% of Customers are missing birth year data
    - 1.52% of Subscribers are missing birth year data  
    - Therefore it is inadvisable to remove all observations with missing birth year data, as this would remove all customers. This will have to be worked around when developing database schema/ views
    
<br>


- All missing birth year observations have 0 recorded for gender (unknown gender)